bson/bsoncodec/struct_codec.go

```diff
@@ -365,3 +392,69 @@ func (sc *StructCodec) describeStruct(r *Registry, t reflect.Type) (*structDescr
 
 	return sd, nil
 }
+
+func fieldByIndexErr(v reflect.Value, index []int) (result reflect.Value, err error) {
+	defer func() {
+		if recovered := recover(); recovered != nil {
+			switch r := recovered.(type) {
+			case string:
+				err = fmt.Errorf("%s", r)
+			case error:
+				err = r
+			}
+		}
+	}()
+
+	result = v.FieldByIndex(index)
+	return
+}
+
+// recursivePointerTo calls reflect.New(v.Type) but recursively for its fields inside
+func recursivePointerTo(v reflect.Value) reflect.Value {
+	v = reflect.Indirect(v)
+	result := reflect.New(v.Type())
+	if v.Kind() == reflect.Struct {
+		for i := 0; i < v.NumField(); i++ {
+			if f := v.Field(i); f.Kind() == reflect.Ptr {
+				if f.Elem().Kind() == reflect.Struct {
+					result.Elem().Field(i).Set(recursivePointerTo(f))
+				}
+			}
+		}
+	}
+
+	return result
+}
+
+// deepZero returns recursive zero object
+func deepZero(st reflect.Type) reflect.Value {
```

Comment:
How is this different than https://golang.org/pkg/reflect/#Zero?

Reply:
Not sure when this was introduced, so might need to care about go version here.

Reply:
removed. Unlike reflect.Zero, it also instantiates all pointers to zeroed structs instead of nil. This will probably come back when I do the mgocompat struct work